tensorflow/contrib/opt/python/training/irprop_plus.py

```diff
@@ -0,0 +1,319 @@
+
+  def _prepare(self):
+    self._eta_minus_t = ops.convert_to_tensor(self._eta_minus, name="eta_minus")
+    self._eta_plus_t = ops.convert_to_tensor(self._eta_plus, name="eta_plus")
+    self._delta_min_t = ops.convert_to_tensor(self._delta_min, name="delta_min")
+    self._delta_max_t = ops.convert_to_tensor(self._delta_max, name="delta_max")
+
+  def _create_slots(self, var_list):
+    # Create slots for the gradient at (t-1) and
+    # the step-size "delta_update".
+    for v in var_list:
+      # Gradient from the previous step
+      self._zeros_slot(v, 'old_grad', self._name)
+      # Init value for delta at step t = 0
+      init_step = math_ops.add(array_ops.zeros(
+          v.get_shape().as_list(), dtype=v.dtype.base_dtype), self._delta_zero)
+      self._get_or_make_slot(v, init_step, "delta_update", self._name)
+
+  # Helper method to check if variable is scalar
+  def _is_scalar(self, tensor):
+    return tensor is not None and \
+            tensor.shape.ndims == 0
+           #tensor.shape.ndims is not None
+
+  # Called by apply_gradients in Optimizer base class
+  def _apply_dense(self, grad, var):
+    old_grad = self.get_slot(var, 'old_grad')
+    delta_update = self.get_slot(var, "delta_update")
+
+    eta_minus_t = math_ops.cast(self._eta_minus_t, var.dtype.base_dtype)
+    eta_plus_t = math_ops.cast(self._eta_plus_t, var.dtype.base_dtype)
+    delta_min_t = math_ops.cast(self._delta_min_t, var.dtype.base_dtype)
+    delta_max_t = math_ops.cast(self._delta_max_t, var.dtype.base_dtype)
+    # iRprop+ error tensors
+    error_t = math_ops.cast(self._error, var.dtype.base_dtype)
+    old_error_t = math_ops.cast(self._old_error, var.dtype.base_dtype)
+
+    return training_ops.apply_i_rprop_plus(
+        var,
+        old_grad,
+        delta_update,
+        eta_minus_t,
+        eta_plus_t,
+        delta_min_t,
+        delta_max_t,
+        error_t,
+        old_error_t,
+        grad, use_locking=self._use_locking).op
+
+  def _resource_apply_dense(self, grad, var):
+    old_grad = self.get_slot(var, "old_grad")
+    delta_update = self.get_slot(var, "delta_update")
+
+    eta_minus_t = math_ops.cast(self._eta_minus_t, var.dtype.base_dtype)
+    eta_plus_t = math_ops.cast(self._eta_plus_t, var.dtype.base_dtype)
+    delta_min_t = math_ops.cast(self._delta_min_t, var.dtype.base_dtype)
+    delta_max_t = math_ops.cast(self._delta_max_t, var.dtype.base_dtype)
+    error_t = math_ops.cast(self._error, var.dtype.base_dtype)
+    old_error_t = math_ops.cast(self._old_error, var.dtype.base_dtype)
+
+    return training_ops.resource_apply_i_rprop_plus(
+        var.handle,
+        old_grad.handle,
+        delta_update.handle,
+        eta_minus_t,
+        eta_plus_t,
+        delta_min_t,
+        delta_max_t,
+        error_t,
+        old_error_t,
+        grad, use_locking=self._use_locking)
+
+
+  def minimize(self, loss, global_step=None, var_list=None,
```

Comment:
I don't think you need to overload minimize

Reply:
The iRprop+ optimizer is dependent on the loss value for the update step. This value is needed for the computation in every iteration. This is the part which makes implementing RProp non-trivial.